### PR TITLE
Prevent unwanted emergency yaw reset due to other sensor errors

### PR DESCRIPTION
**Pull request by priseborough**
Fixes https://github.com/ArduPilot/ardupilot/issues/14253

The reset in both these cases was being triggered internally within the EKF and not by the external EKF fail-safe. This change removes the internal trigger mechanism.

Testing using SITL has been performed to reproduce the GPS and accel vibe error cases identified in the issue. It has verified for both EKF2 and EKF3 that:

A simulated GPS position jump does not trigger a yaw reset
A simulated IMU clipping does not trigger a yaw reset

This change does slightly increase the delay before a flyaway due to bad compass yaw results in a yaw reset because it is now reliant on the external EKF fail-safe mechanism, however the functionality has been tested with a  simulated flyaway induced by setting COMPASS_AUTODEC = 0 and COMPASS_DEC = -3 . This resulted in a reset after a lateral displacement of approximately 10m from the takeoff point, eg:

image

image

### CODE REVIEW

**Comment by priseborough** [comment]
After discussion, we agreed the most practical way forward in the short term is to implement the Copter EKF failsafe yaw reset request logic in ArduPlane and apply it when in VTOL flight modes that require position.

**Comment by priseborough** [comment]
I've brought in the relevant parts of the copter EKF checks and applied them to VTOL flight only. SITL testing has performed in AUTO using "../Tools/autotest/sim_vehicle.py -f quadplane -L KSFO --console --map" and the "Tools/autotest/Generic_Missions/KSFO-VTOL.txt" mission was loaded.

A yaw error was induced by setting COMPASS_AUTODEC = 0 and COMPASS_DEC = -3. The yaw reset activated as expected and the mission continued

2020-05-05 09:21:19 ArduPlane V4.1.0dev (a66bfa38)
2020-05-05 09:21:19 6bd480ce472f43f3b5397361621def52
2020-05-05 09:21:19 Param space used: 479/3840
2020-05-05 09:21:19 RC Protocol: SITL
2020-05-05 09:21:19 New mission
2020-05-05 09:21:19 New rally
2020-05-05 09:21:19 GPS 1: detected as u-blox at 115200 baud
2020-05-05 09:21:21 EKF yaw reset -1.20
2020-05-05 09:21:25 Mission: 2 WP
2020-05-05 09:21:26 EKF2 IMU0 emergency yaw reset - mag sensor stopped
2020-05-05 09:21:26 EKF yaw reset -175.80
2020-05-05 09:21:26 Transition airspeed reached 18.0
2020-05-05 09:21:28 Reached waypoint #2 dist 32m
2020-05-05 09:21:28 Mission: 3 WP
2020-05-05 09:21:31 Transition done

image

**Comment by priseborough** [comment]
Rebased to fix conflict due to added parameters

**Comment by priseborough** [comment]
Rebase has broken the fix on my testing - will investigate.

**Comment by priseborough** [comment]
SITL testing of EKF3 dead reckoning was performed with a Plane flight in TAKEOFF mode with GPS failed once the plane had entered the end of takeoff loiter. Following plots show estimated and truth  position:

Screen Shot 2020-05-09 at 10 53 33 pm

Screen Shot 2020-05-09 at 10 52 11 pm

Growth in 1-sigma velocity and position uncertainty from covariance matrix during dead-reckoning was about half the actual growth in error, indicating that some tuning may  be required:

Screen Shot 2020-05-09 at 10 58 23 pm

File: ArduPlane/ekf_check.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    static_assert(EKF_CHECK_ITERATIONS_MAX >= 7, "EKF_CHECK_ITERATIONS_MAX must be at least 7");
+
+    // exit immediately if ekf has no origin yet - this assumes the origin can never become unset
+    Location temp_loc;
+    if (!ahrs.get_origin(temp_loc)) {
+        return;
+    }
+
+    // return immediately if motors are not armed, or ekf check is disabled
+    if (!plane.arming.is_armed() || quadplane.in_vtol_posvel_mode() || (g2.fs_ekf_thresh <= 0.0f)) {
```
Comment (tridge) [inline ArduPlane/ekf_check.cpp:42]: missing a ! on the quadplane.in_vtol_posvel_mode() check

File: ArduPlane/ekf_check.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+#ifndef EKF_CHECK_WARNING_TIME
+ # define EKF_CHECK_WARNING_TIME            (30*1000)   // warning text messages are sent to ground no more than every 30 seconds
+#endif
+
+////////////////////////////////////////////////////////////////////////////////
+// EKF_check structure
+////////////////////////////////////////////////////////////////////////////////
+static struct {
+    uint8_t fail_count;         // number of iterations ekf or dcm have been out of tolerances
+    uint8_t bad_variance : 1;   // true if ekf should be considered untrusted (fail_count has exceeded EKF_CHECK_ITERATIONS_MAX)
```
Comment (tridge) [inline ArduPlane/ekf_check.cpp:23]: better to just be a bool, same space and faster code

File: ArduPlane/ekf_check.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    // return false immediately if disabled
+    if (g2.fs_ekf_thresh <= 0.0f) {
+        return false;
+    }
+
+    // use EKF to get variance
+    float position_variance, vel_variance, height_variance, tas_variance;
+    Vector3f mag_variance;
+    Vector2f offset;
+    ahrs.get_variances(vel_variance, position_variance, height_variance, mag_variance, tas_variance, offset);
```
Comment (tridge) [inline ArduPlane/ekf_check.cpp:113]: should check bool return of get_variances(). If it is false we may be using DCM. I think we should return false in that case

File: ArduPlane/ekf_check.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+    const float mag_max = fmaxf(fmaxf(mag_variance.x,mag_variance.y),mag_variance.z);
+
+    // return true if two of compass, velocity and position variances are over the threshold OR velocity variance is twice the threshold
+    uint8_t over_thresh_count = 0;
+    if (mag_max >= g2.fs_ekf_thresh) {
+        over_thresh_count++;
+    }
+
+    bool optflow_healthy = false;
```
Comment (tridge) [inline ArduPlane/ekf_check.cpp:123]: this needs to be conditioned on optflow being enabled

Comment (priseborough) [reply ArduPlane/ekf_check.cpp:123]: Thinking about this some more I don't think an optical flow check is useful for plane at this time given the existing use case is terrain height estimation. I will remove the optical flow health conditional logic.

File: ArduPlane/ekf_check.cpp (hunk 3 above)
Comment (priseborough) [reply ArduPlane/ekf_check.cpp:113]: will fix

File: ArduPlane/ekf_check.cpp (hunk 2 above)
Comment (priseborough) [reply ArduPlane/ekf_check.cpp:23]: will fix

File: ArduPlane/ekf_check.cpp (hunk 1 above)
Comment (priseborough) [reply ArduPlane/ekf_check.cpp:42]: will fix

File: ArduPlane/ekf_check.cpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    const float mag_max = fmaxf(fmaxf(mag_variance.x,mag_variance.y),mag_variance.z);
+
+    // return true if two of compass, velocity and position variances are over the threshold OR velocity variance is twice the threshold
+    uint8_t over_thresh_count = 0;
+    if (mag_max >= g2.fs_ekf_thresh) {
+        over_thresh_count++;
+    }
+
+    if ((position_variance >= g2.fs_ekf_thresh && over_thresh_count >= 1) || over_thresh_count >= 2) {
```
[the remark this answers is not in the document]
Comment (priseborough) [reply ArduPlane/ekf_check.cpp:125]: I'm still not entirely happy wth the failsafe so given its intended use is during VTOL operation, I will try to get it back to what copter does, excluding optical flow.

File: ArduPlane/ekf_check.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    float position_variance, vel_variance, height_variance, tas_variance;
+    Vector3f mag_variance;
+    Vector2f offset;
+    if (!ahrs.get_variances(vel_variance, position_variance, height_variance, mag_variance, tas_variance, offset)) {
+        return false;
+    };
+
+    const float mag_max = fmaxf(fmaxf(mag_variance.x,mag_variance.y),mag_variance.z);
+
+    // return true if two of compass, velocity and position variances are over the threshold OR velocity variance is twice the threshold
```
[the remark this answers is not in the document]
Comment (priseborough) [reply ArduPlane/ekf_check.cpp:119]: Will go through all comments in this section and update.

File: ArduPlane/ekf_check.cpp [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+#ifndef EKF_CHECK_WARNING_TIME
+ # define EKF_CHECK_WARNING_TIME            (30*1000)   // warning text messages are sent to ground no more than every 30 seconds
+#endif
+
+////////////////////////////////////////////////////////////////////////////////
+// EKF_check structure
+////////////////////////////////////////////////////////////////////////////////
+static struct {
+    uint8_t fail_count;         // number of iterations ekf or dcm have been out of tolerances
+    bool bad_variance : true;   // true if ekf should be considered untrusted (fail_count has exceeded EKF_CHECK_ITERATIONS_MAX)
```
[the remark this answers is not in the document]
Comment (priseborough) [reply ArduPlane/ekf_check.cpp:23]: will fix

**Comment by priseborough** [comment]
I did a quick SITL quadplane test using 'sim_vehicle.py -j4 -f quadplane --console --map' with the following parameter changes to induce a 180 deg yaw error:

COMPASS_AUTODEC = 0
COMPASS_DEC = 3

The yaw reset and recovery was triggered as expected.

**Pull request merged**